pacsman/pynetdicom_client.py

```diff
@@ -0,0 +1,391 @@
+                negotiation.scp_role = True
+                extended_negotiation_info.append(negotiation)
+
+            with association(ae, self.pacs_url, self.pacs_port,
+                             ext_neg=extended_negotiation_info) as assoc:
+                dataset = Dataset()
+                dataset.SeriesInstanceUID = series_id
+                dataset.QueryRetrieveLevel = 'IMAGE'
+
+                if scp.is_alive():
+                    responses = assoc.send_c_move(dataset, scp.ae_title,
+                                                  query_model='S')
+                else:
+                    raise Exception(f'Storage SCP failed to start for series {series_id}')
+
+                for (status, response) in responses:
+                    logger.debug(status)
+                    logger.debug(response)
+
+                    if status.Status not in status_success_or_pending:
+                        raise Exception(
+                            'Image C-MOVE Failure Response: 0x{0:04x}'.format(
+                                status.Status))
+
+                return series_path if os.path.exists(series_path) else None
+
+
+    def fetch_thumbnail(self, series_id):
+        ae = AE(ae_title=self.client_ae, scu_sop_class=QueryRetrieveSOPClassList)
+
+        with association(ae, self.pacs_url, self.pacs_port) as assoc:
+            # search for image IDs in the series
+            find_dataset = Dataset()
+            find_dataset.SeriesInstanceUID = series_id
+            find_dataset.QueryRetrieveLevel = 'IMAGE'
+            find_dataset.SOPInstanceUID = ''
+            find_response = assoc.send_c_find(find_dataset, query_model='S')
+
+            image_ids = []
+            for (status, result) in find_response:
+                logger.debug(status)
+                logger.debug(result)
+                if status.Status in status_success_or_pending:
+                    if hasattr(result, 'SOPInstanceUID'):
+                        image_ids.append(result.SOPInstanceUID)
+                else:
+                    raise Exception('Thumbnail C-FIND Failure Response: 0x{0:04x}'.format(
+                                    status.Status))
+
+            if not image_ids:
+                return None
+
+
+            with storage_scp(self.client_ae, self.dicom_dir) as scp:
+                # get the middle image in the series for the thumbnail
+                middle_image_id = image_ids[len(image_ids) // 2]
+                move_dataset = Dataset()
+                move_dataset.SOPInstanceUID = middle_image_id
+                move_dataset.QueryRetrieveLevel = 'IMAGE'
+
+                if scp.is_alive():
+                    response = assoc.send_c_move(move_dataset, scp.ae_title,
+                                                 query_model='S')
+                else:
+                    raise Exception(f'Storage SCP failed to start for series {series_id}')
+
+                for (status, d) in response:
```

Comment:
I think my first comment about this wasn't clear.  So it seems that the lines:

```
+                for (status, d) in response:
 +                    logger.debug(status)
 +                    logger.debug(d)
 +                    if status.Status not in status_success_or_pending:
 +                        raise Exception(
 +                            'Thumbnail C-MOVE Failure Response: 0x{0:04x}'.format(
 +                                status.Status))
```

Are repeated a lot.  What if we added a method that does:

```python
def check_responses(responses):
    for (status, dataset) in response:
        logger.debug(status)
        logger.debug(dataset)
        if status.Status in status_success_or_pending:
            yield dataset
        else:
            raise FailedDICOMException('Response: 0x{0:04x}'.format(status.Status))
```

Reply:
That's very nice! Added in latest.